server/src/test/java/org/elasticsearch/search/query/QueryPhaseCollectorManagerTests.java

```diff
@@ -0,0 +1,719 @@
+import org.apache.lucene.tests.index.RandomIndexWriter;
+import org.apache.lucene.tests.search.DummyTotalHitCountCollector;
+import org.elasticsearch.core.IOUtils;
+import org.elasticsearch.test.ESTestCase;
+import org.hamcrest.Matchers;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.List;
+
+public class QueryPhaseCollectorManagerTests extends ESTestCase {
+    private Directory directory;
+    private IndexReader reader;
+    private IndexSearcher searcher;
+    private int numDocs;
+    private int numField2Docs;
+    private int numField3Docs;
+    private int numField2AndField3Docs;
+
+    @Override
+    public void setUp() throws Exception {
+        super.setUp();
+        directory = newDirectory();
+        RandomIndexWriter writer = new RandomIndexWriter(random(), directory, newIndexWriterConfig());
+        numDocs = randomIntBetween(900, 1000);
+        for (int i = 0; i < numDocs; i++) {
+            Document doc = new Document();
+            doc.add(new StringField("field1", "value", Field.Store.NO));
+            boolean field2 = randomBoolean();
+            if (field2) {
+                doc.add(new StringField("field2", "value", Field.Store.NO));
+                numField2Docs++;
+            }
+            boolean field3 = randomBoolean();
+            if (field3) {
+                doc.add(new StringField("field3", "value", Field.Store.NO));
+                numField3Docs++;
+            }
+            if (field2 && field3) {
+                numField2AndField3Docs++;
+            }
+            writer.addDocument(doc);
+        }
+        writer.flush();
+        reader = writer.getReader();
+        searcher = newSearcher(reader);
+        writer.close();
+    }
+
+    @Override
+    public void tearDown() throws Exception {
+        super.tearDown();
+        IOUtils.close(reader, directory);
+    }
+
+    /**
+     * Collector manager used temporarily to bridge tests and query phase as they have different expectations when it comes to their
+     * return type and how results are retrieved from a collector manager.
+     */
+    private static class CollectorManagerAdapter<C extends Collector, T> implements CollectorManager<Collector, Void> {
+        private final CollectorManager<C, T> wrapped;
+        private T result;
+
+        CollectorManagerAdapter(CollectorManager<C, T> wrapped) {
+            this.wrapped = wrapped;
+        }
+
+        @Override
+        public Collector newCollector() throws IOException {
+            return wrapped.newCollector();
+        }
+
+        @Override
+        public Void reduce(Collection<Collector> collectors) throws IOException {
+            @SuppressWarnings("unchecked")
+            List<C> cs = collectors.stream().map(collector -> (C) collector).toList();
+            result = wrapped.reduce(cs);
+            return null;
+        }
+
+        public T getResult() {
+            return result;
+        }
```

Comment:
I relegated the typing and awkward things in this test class for now. This is what it takes to be able to get results of the collection if we don't leverage the return type of the search method. Less generics, but each collector manager needs to hold mutable state that is assigned when reduce is called, and additional getters to retrieve them. In the long run, the query phase will have a well defined set of expected return types that can simplify things.